File: src/main/clojure/comzotohlabscljc/netty/expect100.clj
;; This library is distributed in  the hope that it will be useful but without
;; any  warranty; without  even  the  implied  warranty of  merchantability or
;; fitness for a particular purpose.
;; The use and distribution terms for this software are covered by the Eclipse
;; Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
;; can be found in the file epl-v10.html at the root of this distribution.
;; By using this software in any  fashion, you are agreeing to be bound by the
;; terms of this license. You  must not remove this notice, or any other, from
;; this software.
;; Copyright (c) 2013-2014 Cherimoia, LLC. All rights reserved.


(ns ^{ :doc ""
       :author "kenl" }

  comzotohlabscljc.netty.expect100 )

(use '[clojure.tools.logging :only [info warn error debug] ])

(import '(io.netty.channel ChannelFutureListener ChannelFuture
  ChannelHandler ChannelHandlerContext
  ChannelInboundHandlerAdapter))

(import '(io.netty.buffer ByteBuf ByteBufHolder Unpooled))

(import '(io.netty.handler.codec.http
  HttpMessage HttpResponseStatus
  HttpHeaders HttpVersion
  DefaultFullHttpResponse))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(def ^:private CONTINUE (DefaultFullHttpResponse. (HttpVersion/HTTP_1_1)
                                                  (HttpResponseStatus/CONTINUE)
                                                  (Unpooled/EMPTY_BUFFER)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn HandleExpect100 ""

  ^ChannelHandler
  []

  (proxy [ChannelInboundHandlerAdapter] []
    (channelRead [c msg]
      (let [^ChannelHandlerContext ctx c]
        (when (and (instance?  HttpMessage msg)
                   (HttpHeaders/is100ContinueExpected msg))
          (-> (.writeAndFlush ctx CONTINUE)
              (.addListener (reify ChannelFutureListener
                              (operationComplete [_ ^ChannelFuture f]
                                (when-not (.isSuccess f)
                                          (.fireExceptionCaught ctx (.cause f))))))))
        (.fireChannelRead ctx msg)

      ))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def ^:private expect100-eof nil)

File: src/main/java/com/zotohlab/frwk/netty/Expect100.java
/*??
// This library is distributed in  the hope that it will be useful but without
// any  warranty; without  even  the  implied  warranty of  merchantability or
// fitness for a particular purpose.
// The use and distribution terms for this software are covered by the Eclipse
// Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
// can be found in the file epl-v10.html at the root of this distribution.
// By using this software in any  fashion, you are agreeing to be bound by the
// terms of this license. You  must not remove this notice, or any other, from
// this software.
// Copyright (c) 2013 Cherimoia, LLC. All rights reserved.
 ??*/

package com.zotohlab.frwk.netty;

import io.netty.channel.*;
import io.netty.handler.codec.http.*;
import io.netty.util.ReferenceCountUtil;

/**
 * @author kenl
 */
@ChannelHandler.Sharable
public class Expect100 extends SimpleInboundHandler {

  private static final Expect100 shared = new Expect100();
  public static Expect100 getInstance() {
    return shared;
  }

  public static ChannelPipeline addLast(ChannelPipeline pipe) {
    pipe.addLast(Expect100.class.getSimpleName(), shared);
    return pipe;
  }

  public Expect100() {
  }

  public static void handle100( final ChannelHandlerContext ctx, HttpMessage msg) {
    if (HttpHeaders.is100ContinueExpected(msg)) {
      ctx.writeAndFlush( c100_rsp() ).addListener(new ChannelFutureListener() {
        public void operationComplete(ChannelFuture f) {
          if (!f.isSuccess()) {
            ctx.fireExceptionCaught(f.cause());
          }
        }
      });
    }
  }

  @Override
  protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Exception {
    if (msg instanceof HttpMessage) {
      handle100(ctx, (HttpMessage) msg);
    }
    // simplechannelinboundhandler does a release, so add one here
    ReferenceCountUtil.retain(msg);
    ctx.fireChannelRead(msg);
  }

  private static HttpResponse c100_rsp() {
    return new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,
                                                  HttpResponseStatus.CONTINUE);
  }
}

